OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnLayer.cs

```diff
@@ -0,0 +1,71 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public abstract class GrantConditionOnLayerInfo : ConditionalTraitInfo
+	{
+		[FieldLoader.Require]
+		[GrantedConditionReference]
+		[Desc("The condition to grant to self when changing to specific custom layer.")]
+		public readonly string Condition = null;
+	}
+
+	public abstract class GrantConditionOnLayer : ConditionalTrait<GrantConditionOnLayerInfo>, INotifyChangedCustomLayer
+	{
+		protected ConditionManager conditionManager;
+		protected int conditionToken = ConditionManager.InvalidConditionToken;
+
+		protected virtual byte ValidLayerType { get; set; }
```

Comment:
Can you make this a `protected readonly byte validLayerType` and then assign it from the ctor here?